core/vm/evm.go

```diff
@@ -82,10 +97,14 @@ func run(evm *EVM, contract *Contract, input []byte, readOnly bool) ([]byte, err
 					// Override the input with vrf data of the requested block so it can be returned to the contract program.
 					input = evm.Context.VRF.Bytes()
 				}
-
+			} else if _, ok := p.(*epoch); ok {
+				input = evm.EpochNumber.Bytes()
 			}
 			return RunPrecompiledContract(p, input, contract)
 		}
+		if p := writeCapablePrecompiles[*contract.CodeAddr]; p != nil {
+			return RunWriteCapablePrecompiledContract(p, evm, contract, input, readOnly)
```

Comment:
I forgot the reason. Why do we have to have separate "WriteCapable" Precompiled contract map?

Reply:
nvm. it's because we need to inject evm context.

Reply:
For the extra `evm` and `readOnly` parameters which need to be cascaded to the precompile. The `evm` parameter is the one which handles the staking, and the `readOnly` guarantees that Solidity cannot call the staking precompile via `staticcall`.